go-controller/pkg/clustermanager/userdefinednetwork/controller.go

```diff
@@ -115,25 +124,130 @@ func (c *Controller) UserDefinedNetworkReconciler(key string) error {
+	nadCopy, syncErr := c.SyncUserDefinedNetwork(udnCopy, nadCopy)
 
 	updateStatusErr := c.UpdateUserDefinedNetworkStatus(udnCopy, nadCopy, syncErr)
 
 	return errors.Join(syncErr, updateStatusErr)
 }
 
-func (c *Controller) SyncUserDefinedNetwork(udn *userdefinednetworkv1.UserDefinedNetwork, nad *netv1.NetworkAttachmentDefinition) error {
-	//TODO: implement sync UDN objects logic
-	return errors.New("implement me")
+func (c *Controller) SyncUserDefinedNetwork(udn *userdefinednetworkv1.UserDefinedNetwork, nad *netv1.NetworkAttachmentDefinition) (*netv1.NetworkAttachmentDefinition, error) {
+	if udn == nil {
+		return nil, nil
+	}
+
+	desiredNAD, err := c.nadRenderer.RenderNetAttachDefManifest(udn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate NetworkAttachmetDefinition: %w", err)
+	}
+	if nad == nil {
+		nad, err = c.nadClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(udn.Namespace).Create(context.Background(), desiredNAD, metav1.CreateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to create NetworkAttachmetDefinition: %w", err)
+		}
+		klog.Infof("Created NetworkAttachmentDefinition [%s/%s]", nad.Namespace, nad.Name)
+		return nad, nil
+	}
+
+	if !ownedByUserDefinedNetwork(nad, udnGroupVersionKind, udn.UID) {
+		return nil, fmt.Errorf("foreign NetworkAttachmetDefinition with the desired name already exist")
+	}
+
+	if !reflect.DeepEqual(nad.Spec, desiredNAD.Spec) ||
+		!reflect.DeepEqual(nad.ObjectMeta.OwnerReferences, desiredNAD.ObjectMeta.OwnerReferences) {
+		nad.Spec.Config = desiredNAD.Spec.Config
+		nad.ObjectMeta.OwnerReferences = desiredNAD.ObjectMeta.OwnerReferences
+		nad, err = c.nadClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(nad.Namespace).Update(context.Background(), nad, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to update NetworkAttachmetDefinition: %w", err)
+		}
+		klog.Infof("Updated NetworkAttachmetDefinition [%s/%s]", nad.Namespace, nad.Name)
+	}
+
+	return nad, nil
+}
+
+func ownedByUserDefinedNetwork(nad *netv1.NetworkAttachmentDefinition, gvk schema.GroupVersionKind, uid types.UID) bool {
+	for _, ownerRef := range nad.OwnerReferences {
+		if ownerRef.UID == uid && ownerRef.Kind == gvk.Kind && ownerRef.APIVersion == gvk.GroupVersion().String() {
+			return true
+		}
+	}
+	return false
 }
 
 func (c *Controller) UpdateUserDefinedNetworkStatus(udn *userdefinednetworkv1.UserDefinedNetwork, nad *netv1.NetworkAttachmentDefinition, syncError error) error {
-	//TODO: implement UDN status update logic
-	return errors.New("implement me")
+	if udn == nil {
+		return nil
+	}
+
+	networkReadyCondition := newNetworkReadyCondition(nad, syncError)
+
+	var updatedConditions []metav1.Condition
+	if len(udn.Status.Conditions) == 0 {
+		updatedConditions = append(updatedConditions, *networkReadyCondition)
+	} else {
+		cp := slices.Clone(udn.Status.Conditions)
+		updatedConditions = updateCondition(cp, networkReadyCondition)
+	}
```

Comment:
I think it makes sense to put both cases inside `updateCondition` function, and just append condition when list is empty, then you could also return a bool to signal if the conditions were updated, and you won't need DeepEqual check in the next line

Reply:
I am a bit lost, which additional field do you mean? will it be added in the follow up?